osu.Game.Rulesets.Mania/Difficulty/ManiaPerformanceCalculator.cs

```diff
@@ -55,30 +59,219 @@ protected override PerformanceAttributes CreatePerformanceAttributes(ScoreInfo s
+
+            if (score.Mods.Any(m => m is ModHardRock))
+                windowMultiplier *= 1 / 1.4;
+            else if (score.Mods.Any(m => m is ModEasy))
+                windowMultiplier *= 1.4;
+
+            judgements[0] = Math.Floor(16 * windowMultiplier);
+            judgements[1] = Math.Floor((64 - 3 * overallDifficulty) * windowMultiplier);
+            judgements[2] = Math.Floor((97 - 3 * overallDifficulty) * windowMultiplier);
+            judgements[3] = Math.Floor((127 - 3 * overallDifficulty) * windowMultiplier);
+            judgements[4] = Math.Floor((151 - 3 * overallDifficulty) * windowMultiplier);
+
+            return judgements;
+        }
+
+        private double[] getLazerHitWindows(ScoreInfo score, ManiaDifficultyAttributes attributes)
+        {
+            double[] hitWindows = new double[5];
+
+            var track = new TrackVirtual(10000);
+            score.Mods.OfType<IApplicableToTrack>().ForEach(m => m.ApplyToTrack(track));
+            double clockRate = track.Rate;
+
+            double windowMultiplier = 1 / clockRate;
+
+            if (score.Mods.Any(m => m is ModHardRock))
+                windowMultiplier *= 1 / 1.4;
+            else if (score.Mods.Any(m => m is ModEasy))
+                windowMultiplier *= 1.4;
+
+            if (attributes.OverallDifficulty < 5)
+                hitWindows[0] = (22.4 - 0.6 * attributes.OverallDifficulty) * windowMultiplier;
+            else
+                hitWindows[0] = (24.9 - 1.1 * attributes.OverallDifficulty) * windowMultiplier;
+            hitWindows[1] = (64 - 3 * attributes.OverallDifficulty) * windowMultiplier;
+            hitWindows[2] = (97 - 3 * attributes.OverallDifficulty) * windowMultiplier;
+            hitWindows[3] = (127 - 3 * attributes.OverallDifficulty) * windowMultiplier;
+            hitWindows[4] = (151 - 3 * attributes.OverallDifficulty) * windowMultiplier;
+
+            return hitWindows;
+        }
+
+        private double logPcNote(double x, double deviation) => logErfcApprox(x / (deviation * Math.Sqrt(2)));
+        private double logErfcApprox(double x) => x <= 5 ? Math.Log(SpecialFunctions.Erfc(x)) : -Math.Pow(x, 2) - Math.Log(x) - Math.Log(Math.Sqrt(Math.PI));
+
+        // Legacy LN tails take the absolute error of both hit judgements on an LN, so we use a folded normal distribution to calculate it.
+        private double logPcHitHoldTail(double x, double deviation) => Math.Log(1 - Math.Pow(2 * Normal.CDF(0, deviation * Math.Sqrt(2), x) - 1, 2));
+
+        // Log rules make addition and subtraction of the non-log value non-trivial, these methods simply add and subtract the base value of logs.
+        private double logDiff(double l1, double l2) => l1 + SpecialFunctions.Log1p(-Math.Exp(-(l1 - l2)));
+
+        private double logSum(double l1, double l2)
+        {
+            double maxVal = Math.Max(l1, l2);
+            double minVal = Math.Min(l1, l2);
+
+            if (double.IsNegativeInfinity(maxVal))
```

Comment:
1) Leave a short comment on when a negative infinity occurs
2) Are callers of this function handling the return value `-inf` correctly?

Reply:
This review made me realize a possible NaN in the logDiff function, but after fixing it everything handles -inf perfectly fine.
![image](https://user-images.githubusercontent.com/92956514/219976315-ad615d63-3e6a-454f-a954-a2f853828607.png)